### PR TITLE
daemon, overlord: support for cohort-key in refresh and switch

### PR DESCRIPTION
**Pull request by chipaca**
This implements support for setting the cohort-key of a snap via
refresh or switch.

It also includes a refactor to `snapstate.Update` and `.Switch` that
makes them take an options struct. In fact the majority of the diff is
this refactor.

Right now they are two completely separate structs, but I was
wondering whether instead `snapstate.UpdateOptions` should embed
`snapstate.SwitchOptions`. Maybe it's too "cute", but it feels like it
makes sense, semantically.

Anyway, here goes nothing.

---
Note this includes #6803. This PR is just about the last commit.

### CODE REVIEW

File: daemon/api.go [hunk 1]
```diff
@@ -1132,7 +1132,13 @@ func snapUpdate(inst *snapInstruction, st *state.State) (string, []*state.TaskSe
 		return "", nil, err
 	}
 
-	ts, err := snapstateUpdate(st, inst.Snaps[0], inst.Channel, inst.Revision, inst.userID, flags)
+	ts, err := snapstateUpdate(st, inst.userID, &snapstate.UpdateOptions{
```
Comment (pedronis) [inline daemon/api.go]: it's probably indeed easier to do this in two parts, first the refactor then the new feature.  I think userID should still come toward the end of the signature.

File: daemon/api.go [hunk 2]
```diff
@@ -1132,7 +1132,13 @@ func snapUpdate(inst *snapInstruction, st *state.State) (string, []*state.TaskSe
...
+		Name:      inst.Snaps[0],
```
Comment (pedronis) [inline daemon/api.go]: not sure name belongs in Options themselves

File: daemon/api.go [hunk 3]
```diff
@@ -1132,7 +1132,13 @@ func snapUpdate(inst *snapInstruction, st *state.State) (string, []*state.TaskSe
...
+		Channel:   inst.Channel,
+		Revision:  inst.Revision,
+		CohortKey: inst.CohortKey,
+		Flags:     flags,
```
Comment (pedronis) [inline daemon/api.go]: wondering also if flags should still be carried separately

**Comment by chipaca** [comment]
Closing this as it's getting replaced by #6898 and a follow-up to that one.

**Pull request closed without merging**